// ic-mops/dist/commands/test/reporters/silent-reporter.js
import chalk from 'chalk';
import { absToRel } from '../utils.js';
export class SilentReporter {
    constructor() {
        this.passed = 0;
        this.failed = 0;
        this.skipped = 0;
        this.passedFiles = 0;
        this.failedFiles = 0;
        this.passedNamesFlat = [];
    }
    addFiles(_files) { }
    addRun(file, mmf, state, _wasiMode) {
        state.then(() => {
            this.passed += mmf.passed;
            this.failed += mmf.failed;
            this.skipped += mmf.skipped;
            this.passedNamesFlat = [...this.passedNamesFlat, ...mmf.passedNamesFlat];
            if (mmf.passed === 0 && mmf.failed === 0) {
                this.passed++;
                this.passedNamesFlat.push(absToRel(file));
            }
            this.passedFiles += Number(mmf.failed === 0);
            this.failedFiles += Number(mmf.failed !== 0);
            if (mmf.failed) {
                console.log(chalk.red('✖'), absToRel(file));
                mmf.flush('fail');
                console.log('-'.repeat(50));
            }
        });
    }
    done() {
        return this.failed === 0;
    }
}
