// @ic-reactor/core/dist/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACTOR_INITIAL_STATE = exports.LOCAL_INTERNET_IDENTITY_PROVIDER = exports.IC_INTERNET_IDENTITY_PROVIDER = exports.DEFAULT_IC_DIDJS_ID = exports.DEFAULT_LOCAL_DIDJS_ID = exports.LOCAL_HOST_NETWORK_URI = exports.IC_HOST_NETWORK_URI = exports.LOCAL_HOSTS = exports.REMOTE_HOSTS = void 0;
exports.REMOTE_HOSTS = [".github.dev", ".gitpod.io"];
exports.LOCAL_HOSTS = ["localhost", "127.0.0.1"];
exports.IC_HOST_NETWORK_URI = "https://ic0.app";
exports.LOCAL_HOST_NETWORK_URI = "http://127.0.0.1:4943";
exports.DEFAULT_LOCAL_DIDJS_ID = "bd3sg-teaaa-aaaaa-qaaba-cai";
exports.DEFAULT_IC_DIDJS_ID = "a4gq6-oaaaa-aaaab-qaa4q-cai";
exports.IC_INTERNET_IDENTITY_PROVIDER = "https://identity.ic0.app/#authorize";
exports.LOCAL_INTERNET_IDENTITY_PROVIDER = "http://rdmx6-jaaaa-aaaaa-aaadq-cai.localhost:4943/#authorize";
exports.ACTOR_INITIAL_STATE = {
    name: "",
    methodState: {},
    initializing: false,
    initialized: false,
    error: undefined,
};
